exercises/collatz-conjecture/collatz_conjecture_test.go

```diff
@@ -0,0 +1,77 @@
+package collatzconjecture
+
+import (
+	"testing"
+)
+
+var testCases = []struct {
+	description string
+	input       int
+	expectError bool
+	expected    int
+}{
+	{
+		description: "zero steps for one",
+		input:       1,
+		expectError: false,
+		expected:    0,
+	},
+	{
+		description: "divide if even",
+		input:       16,
+		expectError: false,
+		expected:    4,
+	},
+	{
+		description: "even and old steps",
+		input:       12,
+		expectError: false,
+		expected:    9,
+	},
+	{
+		description: "large number of even and odd steps",
+		input:       1000000,
+		expectError: false,
+		expected:    152,
+	},
+	{
+		description: "zero is an error",
+		input:       0,
+		expectError: true,
+	},
+	{
+		description: "negative value is an error",
+		input:       -15,
+		expectError: true,
+	},
+}
+
+func TestCollatzConjecture(t *testing.T) {
+	for _, testCase := range testCases {
+		steps, err := CollatzConjecture(testCase.input)
+		if testCase.expectError {
+			if err == nil {
+				t.Fatalf("FAIL: %s\n\tCollatzConjecture(%v) expected an error, got %v",
+					testCase.description, testCase.input, steps)
+			}
+		} else {
+			if steps != testCase.expected {
+				t.Fatalf("FAIL: %s\n\tCollatzConjecture(%v) expected %v, got %v",
+					testCase.description, testCase.input, testCase.expected, steps)
+			}
+			if err != nil {
```

Comment:
Suggest moving the test for err != nil before the test on steps != testCase.expected.
The reason is that one should test for err before the normal return value. Generally in Go, the return value does not necessarily have a useful value if an error is returned.  It may seem a bit strict here, but it is good to establish that way of thinking about testing for error, and for the benefit of anyone reading the test code as well.